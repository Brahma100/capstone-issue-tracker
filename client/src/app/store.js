import { configureStore } from '@reduxjs/toolkit';
import userReducer from '../features/user/userSlice';
import issueReducer from '../features/issue/issueSlice';

export const store = configureStore({
  reducer: {
  
    user:userReducer,
    issue:issueReducer,
  },
});
