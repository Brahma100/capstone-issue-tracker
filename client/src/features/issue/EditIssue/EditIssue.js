import React,{useEffect, useState} from 'react';
import {Col, Card, Button, Row,Form,Container, InputGroup,Spinner, Tooltip} from 'react-bootstrap';
import {Field, Formik} from 'formik';
import * as yup from 'yup';
import {useDispatch,useSelector} from 'react-redux';
import {selectAuth, selectUser} from '../../user/userSlice';
import { editIssueAsync } from '../issueSlice';
import { Prompt } from 'react-router';


const schemaPro = yup.object({
    Issue:yup.string().required("Issue Name is Required"),
    Severity:yup.string().required(),
    Status:yup.string().required("Status  is Required"),
    Description:yup.string().required("Description  is Required"),
    Comments:yup.string().required("Comments  is Required"),

})


function EditIssue(props){
  const issue=props.history.location.state.issue;
const dispatch=useDispatch();
const userinfo= useSelector(selectUser);
  const [msg,setMsg]=useState(null);
  const [isBlocking,setIsBlocking]=useState(false);
  const [Id,setId]=useState(issue.id); 
  const [Issue,setIssue]=useState(issue.Issue); 
  const [Description,setDescription]=useState(issue.Description); 
  const [Severity,setSeverity]=useState(issue.Severity); 
  const [Status,setStatus]=useState(issue.Status); 
  const [Comments,setComments]=useState(''); 
  const isAuthenticated=useSelector(selectAuth);





useEffect(()=>{
      
  if(!isAuthenticated){
      // console.log("From Product to Home");
      props.history.push('/')
    
     }

},[props,isAuthenticated]);


    var isLoaded= isLoaded;
   
    return (
      <div className="content">
        <Container fluid>
          <Prompt
                when={ isBlocking}
                message={(location)=> `Are You Sure Want To Go To ${location.pathname}`}
/> 
          <Row>
            <Col md={6}>
<Card style={{display:'flex',justifyContent:'space-between'}}>


  <Card.Header>
    <p style={{fontSize:'1.5rem',fontWeight:'bold',color:'#3b44c1'}}>Edit the Choosen Issue</p>
  </Card.Header>    


  <Card.Body>
  <Formik
  validationSchema={schemaPro}
 

  initialValues={{Issue:Issue,Description:Description,Severity:Severity,Status:Status,Comments:Comments}}

  onSubmit={(values)=>{ 

    const {Issue,Description,Severity,Status,Comments}=values;
    // console.log("On Submit Called",Issue);
    
    const user= userinfo;
    // console.log("User......",user);
    const id=Id;
    const issue={
        id,Issue,Description,Severity,Status,user,Comments
    }
    console.log("Edit Data",issue);
    // console.log("Submitted Issue:.......",issue)
  //  console.log("Name:",name," Des:",description," Manu:",manufacturer," price:",price," Stock:",stock," Img:",img,"  Cat",CategoryName," User:",user);
     dispatch(editIssueAsync(issue));
     setIsBlocking(false);
    //  console.log("IsBlocking",isBlocking);
      props.history.goBack();
   
   }
 }
>
  {({
    handleSubmit,
    handleChange,
    values,
    errors,
    touched
     }) => (
    <Form noValidate onSubmit={handleSubmit}>
      
      
      <Form.Row>
      <Form.Group as={Col} md="12" controlId="validationFormik01" style={{width:'92%',paddingLeft:'2rem'}}>
      
          <Form.Label>Issue Name</Form.Label>
          
            <Form.Control
              type="text"
              placeholder="Issue Name"
              aria-describedby="inputGroupPrepend"
              name="Issue"
              value={values.Issue}
              onChange={(e)=>{ setIsBlocking(e.target.value.length>0);handleChange(e); setIssue(e.target.value)}} 
              isInvalid={!!errors.Issue}
            />
            <Form.Control.Feedback type="invalid">
              {errors.Issue}
            </Form.Control.Feedback>
          
        </Form.Group>
        </Form.Row>
        <Form.Group as={Col} md="12" controlId="validationFormik01" style={{width:'92%',paddingLeft:'2rem'}}>
                <Form.Row>
                      <Form.Label>Description</Form.Label>
          
                <textarea 
                  type="text-area"
                  placeholder="Type Description"
                  aria-describedby="inputGroupPrepend"
                  name="Description"
                  value={values.Description}
                  onChange={(e)=>{ setIsBlocking(e.target.value.length>0);handleChange(e); setDescription(e.target.value)}} 
                  isInvalid={!!errors.Description}
                  />
                   {errors.Description && touched.Description && (
                        <div className="input-feedback">{errors.Description}</div>
                    )}
                  </Form.Row> 
        </Form.Group>



        <Form.Row>
      <Form.Group as={Col} md="12" controlId="validationFormik01" style={{width:'92%',paddingLeft:'2rem'}}>

        <Form.Label>Severity</Form.Label>
                    {/* <input className={errors.quantity && touched.quantity && "error"} value={values.quantity} onChange={handleChange} onBlur={handleBlur} type="number" placeholder="Enter Product Quantity" name="quantity"/><br/> */}
                    <select name="Severity" id="Severity" value={values.Severity}  onChange={handleChange}>
                    <option value="Critical">Critical</option>
                    <option value="Minor">Minor</option>
                    <option value="Major">Major</option>
                    
                    </select>
                   
                    {errors.Severity && touched.Severity && (
                        <div className="input-feedback">{errors.Severity}</div>
                    )}
                    <br/>
                   
                    </Form.Group>
        </Form.Row>
        <Form.Row>
      <Form.Group  as={Col} md="12" controlId="validationFormik01" style={{width:'92%',paddingLeft:'2rem'}}>

                    <div value={values.Status} onChange={handleChange} role="group" aria-labelledby="my-radio-group" >
                    <Form.Label>Status</Form.Label>
                        <label className="group">
                        <Field type="radio" name="Status" value="Open" />
                        Open
                        </label>
                        <label className="group">
                        <Field type="radio" name="Status" value="In Progress" />
                        In Progress
                        </label>
                        <label className="group">
                        <Field type="radio" name="Status" value="Closed" />
                        Closed
                        </label>
                       

</div>
 
                    {errors.Status && touched.Status && (
                        <div className="input-feedback">{errors.Status}</div>
                    )}


</Form.Group>
        </Form.Row>
        <Form.Group as={Col} md="12" controlId="validationFormik01" style={{width:'92%',paddingLeft:'2rem'}}>
                <Form.Row>
                      <Form.Label>Comments</Form.Label>
          
                <textarea 
                  type="text-area"
                  placeholder="Type Comments"
                  aria-describedby="inputGroupPrepend"
                  name="Comments"
                  value={values.Comments}
                  onChange={(e)=>{ setIsBlocking(e.target.value.length>0);handleChange(e); setComments(e.target.value)}} 
                  isInvalid={!!errors.Comments}
                  />
                   {errors.Comments && touched.Comments && (
                        <div className="input-feedback">{errors.Comments}</div>
                    )}
                  </Form.Row> 
        </Form.Group>

        
      <Form.Group style={{width:'92%',paddingLeft:'2rem'}}>
      <Button style={{marginTop:'10px'}} type="submit">Submit</Button>
      </Form.Group>
    </Form>
  )}
</Formik>
</Card.Body>
</Card>

                        </Col>
     
          </Row>
        </Container>
      </div>
    );
  }



export default EditIssue;
