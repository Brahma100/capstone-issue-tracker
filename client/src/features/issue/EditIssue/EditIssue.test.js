import { fireEvent,} from "@testing-library/react";
import React from 'react';
import {
  render,
  screen,
} from '@testing-library/react';
import { Provider } from 'react-redux';
import { store } from '../../../app/store';
import  EditIssue from './EditIssue';
import { MemoryRouter } from 'react-router-dom';


const historyMock = {

    location:{ state:{
        issue:{"id":"2827854021bc7a3e7cb73fee","Issue":" Ineffective data edits","Description":"an alphanumeric address field that allows spaces to be entered before any numbers or letters in the address. Therefore, when searches or sorts are performed on the address field, the search or sort may not find the intended address.","Severity":"Minor","Status":"Open","user":{"_id":10,"fname":"Raj","lname":"Jai","email":"1@a.a","city":"Kolkata","state":"West Bengal","country":"India","postal":"700059","password":"$2a$10$KpgATnkGAu.zrRMEoSMFW.8scjfTBN1w46sfGT47wpzh9nR2zr9ly","ip":"103.240.99.25","date":"2021-05-21 22:09:34"},"date":"2021-05-27 17:44:50","rank":0,"editUser":[],"r_date":""}
    }
    },
    push:jest.fn(),
    // goBack:jest.fn()
    };


const renderIssue = () =>
  render(
    <Provider  store={store}>
        < MemoryRouter>
      <EditIssue history={historyMock}/>
      </ MemoryRouter>
    </Provider>,
  );

describe('EditIssue Render',()=>{
 
  test('Checking Form Header', () => {
    const { asFragment } = renderIssue();
    expect(asFragment()).toMatchSnapshot();
    expect(screen.getByText('Edit the Choosen Issue')).toBeInTheDocument();

  });
  test('checking Form Avaiability', () => {
    const { asFragment } = renderIssue();
    expect(asFragment()).toMatchSnapshot();
    const ele=screen.getByText((content, element) => {
        return element.tagName.toLowerCase() === 'form'
      })
    expect(ele).toBeInTheDocument();

  });
  test('Checking the Passed Description Data && Submit Button', () => {
    const { rerender,asFragment } = renderIssue();
    const button =screen.getByRole('button',{name:/submit/i});
    expect(button).toBeInTheDocument();
    expect(screen.getByText("an alphanumeric address field that allows spaces to be entered before any numbers or letters in the address. Therefore, when searches or sorts are performed on the address field, the search or sort may not find the intended address.")).toBeInTheDocument();
    
});
test('Verifying the Passed Comments To the Comments TextArea', () => {
    const { rerender,asFragment } = renderIssue();
    fireEvent.change(screen.getByPlaceholderText(/type comments/i), { target: { value: 'We will resolve this Issue asap..' } })
    expect(screen.getByText("We will resolve this Issue asap..")).toBeInTheDocument();
 
});
test('Trying To click on Submit Button After filling Comments', () => {
    const { rerender,asFragment } = renderIssue();
    fireEvent.change(screen.getByPlaceholderText(/type comments/i), { target: { value: 'We will resolve this Issue asap..' } })
    expect(screen.getByText("We will resolve this Issue asap..")).toBeInTheDocument();
    const submitButton =screen.getByRole('button',{name:/submit/i});
    fireEvent.click(submitButton);
    // expect(screen.getByText("We will resolve this Issue asap..")).toBeInTheDocument();

});

})
 
  





















// import React from 'react'
// import {shallow} from 'enzyme'
// import toJson from 'enzyme-to-json'
// import EditIssue from './EditIssue'

// jest.mock('react-router-dom', () => ({
//     useLocation: jest.fn().mockReturnValue({
//       pathname: '/another-route',
//       search: '',
//       hash: '',
//       state: null,
//       key: '5nvxpbdafa',
//     }),
//     history:{
//         location:'test'
//     }
// }));

// describe('Test UserForm using Shallow rendering', () => {

// let wrapper;
// let props;

// const historyMock = {

// location:{ state:{
//     issue:{"id":"2827854021bc7a3e7cb73fee","Issue":" Ineffective data edits","Description":"an alphanumeric address field that allows spaces to be entered before any numbers or letters in the address. Therefore, when searches or sorts are performed on the address field, the search or sort may not find the intended address.","Severity":"Minor","Status":"Open","user":{"_id":10,"fname":"Raj","lname":"Jai","email":"1@a.a","city":"Kolkata","state":"West Bengal","country":"India","postal":"700059","password":"$2a$10$KpgATnkGAu.zrRMEoSMFW.8scjfTBN1w46sfGT47wpzh9nR2zr9ly","ip":"103.240.99.25","date":"2021-05-21 22:09:34"},"date":"2021-05-27 17:44:50","rank":0,"editUser":[],"r_date":""}
// }
// }
// };

// beforeEach(() => {
//     props = {
        
//             location:{
//                 state:{
//                     issue:{"id":"2827854021bc7a3e7cb73fee","Issue":" Ineffective data edits","Description":"an alphanumeric address field that allows spaces to be entered before any numbers or letters in the address. Therefore, when searches or sorts are performed on the address field, the search or sort may not find the intended address.","Severity":"Minor","Status":"Open","user":{"_id":10,"fname":"Raj","lname":"Jai","email":"1@a.a","city":"Kolkata","state":"West Bengal","country":"India","postal":"700059","password":"$2a$10$KpgATnkGAu.zrRMEoSMFW.8scjfTBN1w46sfGT47wpzh9nR2zr9ly","ip":"103.240.99.25","date":"2021-05-21 22:09:34"},"date":"2021-05-27 17:44:50","rank":0,"editUser":[],"r_date":""}
//                 }
            
//         }
//     }
//     wrapper = shallow(<EditIssue history={historyMock} />);
// })

//     it('renders "Issue Name" as heading in second column' , () => {
//         let secondColumnHeading = wrapper.find('th').at(1);
//         expect(secondColumnHeading.render().text()).toContain('Issue Name');
//      })

//     //  it('renders "Status" as heading in fourth column' , () => {
//     //     let fourthColumnHeading = wrapper.find('th').at(3);
//     //     expect(fourthColumnHeading.render().text()).toContain('Status');
//     //  })
//     //  it("passes the 2nd Issue's name as props to second Issue Component", () => {
//     //    let secondIssue = wrapper.find('Issue').at(1)
//     //     expect(secondIssue.render().text()).toContain('The heading Add is wrongly displayed as Edit');
//     // });

//     // it("passes the 1st Issue's NAME as props to first issue Component", () => {
//     //     let secondIssue = wrapper.find('Issue').at(0)
//     //      expect(secondIssue.render().text()).toContain("On Cliking Delete");
//     //  });
// });


// // describe('When Issues array passed to IssueList component is null', () => {
// //     let wrapper;
// //     let props;

// //     beforeEach(() => {  
// //         props = {
// //             Issues: null 
// //         }
// //         wrapper = shallow(<EditIssue history={historyMock}/>);
// //     }); 

// //     it('should not crash and no Issue Component is rendered', () => {
// //         let Issueelement = wrapper.find('Issue');
// //         expect(Issueelement.length).toEqual(0);
// //     });

    

// // });