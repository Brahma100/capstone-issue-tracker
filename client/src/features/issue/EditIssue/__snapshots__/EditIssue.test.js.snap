// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditIssue Render Checking Form Header 1`] = `
<DocumentFragment>
  <div
    class="content"
  >
    <div
      class="container-fluid"
    >
      <div
        class="row"
      >
        <div
          class="col-md-6"
        >
          <div
            class="card"
            style="display: flex; justify-content: space-between;"
          >
            <div
              class="card-header"
            >
              <p
                style="font-size: 1.5rem; font-weight: bold; color: rgb(59, 68, 193);"
              >
                Edit the Choosen Issue
              </p>
            </div>
            <div
              class="card-body"
            >
              <form
                class=""
                novalidate=""
              >
                <div
                  class="form-row"
                >
                  <div
                    class="form-group col-md-12"
                    style="width: 92%; padding-left: 2rem;"
                  >
                    <label
                      class="form-label"
                      for="validationFormik01"
                    >
                      Issue Name
                    </label>
                    <input
                      aria-describedby="inputGroupPrepend"
                      class="form-control"
                      id="validationFormik01"
                      name="Issue"
                      placeholder="Issue Name"
                      type="text"
                      value=" Ineffective data edits"
                    />
                    <div
                      class="invalid-feedback"
                    />
                  </div>
                </div>
                <div
                  class="form-group col-md-12"
                  style="width: 92%; padding-left: 2rem;"
                >
                  <div
                    class="form-row"
                  >
                    <label
                      class="form-label"
                      for="validationFormik01"
                    >
                      Description
                    </label>
                    <textarea
                      aria-describedby="inputGroupPrepend"
                      name="Description"
                      placeholder="Type Description"
                      type="text-area"
                    >
                      an alphanumeric address field that allows spaces to be entered before any numbers or letters in the address. Therefore, when searches or sorts are performed on the address field, the search or sort may not find the intended address.
                    </textarea>
                  </div>
                </div>
                <div
                  class="form-row"
                >
                  <div
                    class="form-group col-md-12"
                    style="width: 92%; padding-left: 2rem;"
                  >
                    <label
                      class="form-label"
                      for="validationFormik01"
                    >
                      Severity
                    </label>
                    <select
                      id="Severity"
                      name="Severity"
                    >
                      <option
                        value="Critical"
                      >
                        Critical
                      </option>
                      <option
                        value="Minor"
                      >
                        Minor
                      </option>
                      <option
                        value="Major"
                      >
                        Major
                      </option>
                    </select>
                    <br />
                  </div>
                </div>
                <div
                  class="form-row"
                >
                  <div
                    class="form-group col-md-12"
                    style="width: 92%; padding-left: 2rem;"
                  >
                    <div
                      aria-labelledby="my-radio-group"
                      role="group"
                      value="Open"
                    >
                      <label
                        class="form-label"
                        for="validationFormik01"
                      >
                        Status
                      </label>
                      <label
                        class="group"
                      >
                        <input
                          checked=""
                          name="Status"
                          type="radio"
                          value="Open"
                        />
                        Open
                      </label>
                      <label
                        class="group"
                      >
                        <input
                          name="Status"
                          type="radio"
                          value="In Progress"
                        />
                        In Progress
                      </label>
                      <label
                        class="group"
                      >
                        <input
                          name="Status"
                          type="radio"
                          value="Closed"
                        />
                        Closed
                      </label>
                    </div>
                  </div>
                </div>
                <div
                  class="form-group col-md-12"
                  style="width: 92%; padding-left: 2rem;"
                >
                  <div
                    class="form-row"
                  >
                    <label
                      class="form-label"
                      for="validationFormik01"
                    >
                      Comments
                    </label>
                    <textarea
                      aria-describedby="inputGroupPrepend"
                      name="Comments"
                      placeholder="Type Comments"
                      type="text-area"
                    />
                  </div>
                </div>
                <div
                  class="form-group"
                  style="width: 92%; padding-left: 2rem;"
                >
                  <button
                    class="btn btn-primary"
                    style="margin-top: 10px;"
                    type="submit"
                  >
                    Submit
                  </button>
                </div>
              </form>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`EditIssue Render checking Form Avaiability 1`] = `
<DocumentFragment>
  <div
    class="content"
  >
    <div
      class="container-fluid"
    >
      <div
        class="row"
      >
        <div
          class="col-md-6"
        >
          <div
            class="card"
            style="display: flex; justify-content: space-between;"
          >
            <div
              class="card-header"
            >
              <p
                style="font-size: 1.5rem; font-weight: bold; color: rgb(59, 68, 193);"
              >
                Edit the Choosen Issue
              </p>
            </div>
            <div
              class="card-body"
            >
              <form
                class=""
                novalidate=""
              >
                <div
                  class="form-row"
                >
                  <div
                    class="form-group col-md-12"
                    style="width: 92%; padding-left: 2rem;"
                  >
                    <label
                      class="form-label"
                      for="validationFormik01"
                    >
                      Issue Name
                    </label>
                    <input
                      aria-describedby="inputGroupPrepend"
                      class="form-control"
                      id="validationFormik01"
                      name="Issue"
                      placeholder="Issue Name"
                      type="text"
                      value=" Ineffective data edits"
                    />
                    <div
                      class="invalid-feedback"
                    />
                  </div>
                </div>
                <div
                  class="form-group col-md-12"
                  style="width: 92%; padding-left: 2rem;"
                >
                  <div
                    class="form-row"
                  >
                    <label
                      class="form-label"
                      for="validationFormik01"
                    >
                      Description
                    </label>
                    <textarea
                      aria-describedby="inputGroupPrepend"
                      name="Description"
                      placeholder="Type Description"
                      type="text-area"
                    >
                      an alphanumeric address field that allows spaces to be entered before any numbers or letters in the address. Therefore, when searches or sorts are performed on the address field, the search or sort may not find the intended address.
                    </textarea>
                  </div>
                </div>
                <div
                  class="form-row"
                >
                  <div
                    class="form-group col-md-12"
                    style="width: 92%; padding-left: 2rem;"
                  >
                    <label
                      class="form-label"
                      for="validationFormik01"
                    >
                      Severity
                    </label>
                    <select
                      id="Severity"
                      name="Severity"
                    >
                      <option
                        value="Critical"
                      >
                        Critical
                      </option>
                      <option
                        value="Minor"
                      >
                        Minor
                      </option>
                      <option
                        value="Major"
                      >
                        Major
                      </option>
                    </select>
                    <br />
                  </div>
                </div>
                <div
                  class="form-row"
                >
                  <div
                    class="form-group col-md-12"
                    style="width: 92%; padding-left: 2rem;"
                  >
                    <div
                      aria-labelledby="my-radio-group"
                      role="group"
                      value="Open"
                    >
                      <label
                        class="form-label"
                        for="validationFormik01"
                      >
                        Status
                      </label>
                      <label
                        class="group"
                      >
                        <input
                          checked=""
                          name="Status"
                          type="radio"
                          value="Open"
                        />
                        Open
                      </label>
                      <label
                        class="group"
                      >
                        <input
                          name="Status"
                          type="radio"
                          value="In Progress"
                        />
                        In Progress
                      </label>
                      <label
                        class="group"
                      >
                        <input
                          name="Status"
                          type="radio"
                          value="Closed"
                        />
                        Closed
                      </label>
                    </div>
                  </div>
                </div>
                <div
                  class="form-group col-md-12"
                  style="width: 92%; padding-left: 2rem;"
                >
                  <div
                    class="form-row"
                  >
                    <label
                      class="form-label"
                      for="validationFormik01"
                    >
                      Comments
                    </label>
                    <textarea
                      aria-describedby="inputGroupPrepend"
                      name="Comments"
                      placeholder="Type Comments"
                      type="text-area"
                    />
                  </div>
                </div>
                <div
                  class="form-group"
                  style="width: 92%; padding-left: 2rem;"
                >
                  <button
                    class="btn btn-primary"
                    style="margin-top: 10px;"
                    type="submit"
                  >
                    Submit
                  </button>
                </div>
              </form>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
